### ISSUE TITLE
Building c++ api on Debian  fails to link.

### ISSUE THREAD

**Issue by rabraker** [OpenFAST/openfast#305]
**Bug description**

Building open fast with the c++ api enabled fails.

**To Reproduce**

1. Compile with 
```
mkdir build && cd build
cmake -DBUILD_OPENFAST_CPP_API:BOOL=ON  ..
make VERBOSE=1
```
produces a link error (see below).

**Expected behavior**
The build should complete without error.

**Screenshots**
During the link phase, here is the terminal output around the failed command:
```
[100%] Linking CXX executable openfastcpp
cd /home/arnold/local/src/openfast/build/glue-codes/openfast-cpp && /usr/local/bin/cmake -E cmake_link_script CMakeFiles/openfastcpp.dir/link.txt --verbose=1
/usr/bin/c++   -fpic -O3 -DNDEBUG  -rdynamic -L/usr/lib -pthread CMakeFiles/openfastcpp.dir/src/FAST_Prog.cpp.o  -o openfastcpp -Wl,-rpath,/usr/lib/x86_64-linux-gnu/openmpi/lib:/usr/lib/x86_64-linux-gnu/hdf5/serial:/usr/local/lib: libopenfastcpplib.a ../../modules-local/openfast-library/libopenfastlib.a /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so -Wl,-Bstatic -lyaml-cpp -Wl,-Bdynamic /usr/lib/x86_64-linux-gnu/hdf5/serial/libhdf5.so -lpthread -lsz -lz -ldl -lm /usr/lib/x86_64-linux-gnu/hdf5/serial/libhdf5_hl.so -lz -lxml2 -ldl ../../modules-local/openfast-library/libopenfast_postlib.a ../../modules-local/supercontroller/libscfastlib.a ../../modules-local/openfoam/libfoamfastlib.a ../../modules-local/openfast-library/libopenfast_prelib.a ../../modules-local/aerodyn/libaerodynlib.a ../../modules-local/aerodyn14/libaerodyn14lib.a ../../modules-local/inflowwind/libifwlib.a ../../modules-local/elastodyn/libelastodynlib.a ../../modules-local/servodyn/libservodynlib.a ../../modules-local/beamdyn/libbeamdynlib.a ../../modules-local/subdyn/libsubdynlib.a ../../modules-local/hydrodyn/libhydrodynlib.a ../../modules-local/orcaflex-interface/liborcaflexlib.a ../../modules-local/extptfm/libextptfm_mckflib.a ../../modules-local/supercontroller/libsctypeslib.a ../../modules-ext/feamooring/libfeamlib.a ../../modules-ext/moordyn/libmoordynlib.a ../../modules-ext/icedyn/libicedynlib.a ../../modules-ext/icefloe/libicefloelib.a ../../modules-ext/map/libmaplib.a ../../modules-ext/map/libmapcpplib.a ../../modules-local/openfoam/libopenfoamtypeslib.a ../../modules-local/nwtc-library/libnwtclibs.a /usr/local/lib/libopenblas.so -ldl -lm /usr/lib/x86_64-linux-gnu/hdf5/serial/libhdf5_hl.so -lxml2 -ldl -lgfortran -lquadmath 
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libyaml-cpp.a(convert.cpp.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libyaml-cpp.a(memory.cpp.o): relocation R_X86_64_32S against symbol `_ZN5boost6detail15sp_counted_base7destroyEv' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libyaml-cpp.a(node_data.cpp.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libyaml-cpp.a(parse.cpp.o): relocation R_X86_64_32 against `.rodata.str1.8' can not be used when making a shared object; recompile with -fPIC
...
...
error: ld returned 1 exit status
glue-codes/openfast-cpp/CMakeFiles/openfastcpp.dir/build.make:125: recipe for target 'glue-codes/openfast-cpp/openfastcpp' failed
```
**OpenFAST Version**
I'm on the github master branch, at commit id f147b80521eff90c19f065eabeceac13de39ac59

**System Information (please complete the following information):**
 
 - OS: Debian 9
 - Compiler:  gcc-6.3, g++-6.3, gfortran-6.3
 - Compiler settings: 
```
>>> cmake --version
cmake version 3.12.0-rc1
```
```
g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/6/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 6.3.0-18+deb9u1' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-6 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-6-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-6-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-6-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1)
```

**Additional context**

The issue is that [cmake/FindYAMLCPP.cmake](https://github.com/OpenFAST/openfast/blob/f147b80521eff90c19f065eabeceac13de39ac59/cmake/FindYAMLCPP.cmake#L21) is set to look for the archive version of `yaml-cpp` first:
```
find_library(YAML_LIBRARIES
NAMES libyaml-cpp.a yaml-cpp
```
which is not compiled with `-fPIC` (it is from the debian repo). But gcc is configured with `--enable-default-pie`, so it is trying to create a position independent executable, but can't because it's linking against the non-PIC `libyaml-cpp.a`.

The simplest fix is to tell cmake to look for the archive second:
```
diff --git a/cmake/FindYAMLCPP.cmake b/cmake/FindYAMLCPP.cmake
index a5fd6db7..a3557319 100644
--- a/cmake/FindYAMLCPP.cmake
+++ b/cmake/FindYAMLCPP.cmake
@@ -18,7 +18,7 @@ find_path(YAML_INCLUDES
   PATH_SUFFIXES include)
 
 find_library(YAML_LIBRARIES
-  NAMES libyaml-cpp.a yaml-cpp
+  NAMES yaml-cpp libyaml-cpp.a
   HINTS ${YAML_ROOT} ${CMAKE_INSTALL_PREFIX}
   PATH_SUFFIXES lib)
```
This works for me.

Alternatives might be to:
1. Specify `-no-pie` while compiling the executables
2. Document the issue in the build instructions so the user can specify `-no-pie` via `CXXFLAGS` etc.
3. ?

**Comment by sayerhs** [comment]
@rafmudaf The proposed fix to FindYAMLCPP.cmake is the right one. We should update the dev branch with it.

**Comment by rafmudaf** [comment]
This should also be turned on in the TravisCI build. I'm working on configuring that now to be sure that the bug is exposed (see https://github.com/rafmudaf/openfast/tree/bugfix/cpp_api).

**Comment by rafmudaf** [comment]
closed by #313
